### ISSUE TITLE
Auth tokens can become invalid on renewal if request is taking long time

### ISSUE THREAD

**Issue by nARN** [grafana/grafana#18644]
**What happened**:
When it is time to renew the auth cookie Grafana frequently logs users out when they look at dashboards issuing heavy long queries. 

**What you expected to happen**:
Cookies should stay valid

**How to reproduce it (as minimally and precisely as possible)**:
As this is sort of a race condition a guaranteed reproduction is somewhat complicated. So here's how it happens in real world. You need:
  - A naturally slow data source which takes more than a minute (better two) to response
  - Short dashboard refresh interval so that queries overlap
  - A dashboard with charts from that data source (not tables)

Eventually, you will be logged out.

I tried to come up with a more reliable reproduction based on postgres where I can artificially slow queries down by locking tables, but it quickly became unnecessarily complicated and fragile. If necessary I can try to finish that, but have a look below at what I think happens.

**Anything else we need to know?**:
A freshly logged in user with Token1 starts watching a dashboard that issues long running queries.
When it is [time to renew the auth token](https://github.com/grafana/grafana/blob/299a0e20f45fd2829fc0aaad44facec9e2e74723/pkg/services/auth/auth_token.go#L206):
  - A new Token2 is generated and [marked as unseen](https://github.com/grafana/grafana/blob/299a0e20f45fd2829fc0aaad44facec9e2e74723/pkg/services/auth/auth_token.go#L233)
  - Token1 is moved to prev_auth_token
  - Token2 is [stored in the response](https://github.com/grafana/grafana/blob/master/pkg/middleware/middleware.go#L232)
  - Query processing begins (and will take a while because the data source is slow)
  - A new request is issued more than [a minute](https://github.com/grafana/grafana/blob/299a0e20f45fd2829fc0aaad44facec9e2e74723/pkg/services/auth/auth_token.go#L26) after the previous one but before it finishes.
  - The new token has not been sent to the UI yet as the query is still running. So the new request still has Token1.
  - Grafana [accepts Token1](https://github.com/grafana/grafana/blob/299a0e20f45fd2829fc0aaad44facec9e2e74723/pkg/services/auth/auth_token.go#L112) and decides to [rotate the token again](https://github.com/grafana/grafana/blob/299a0e20f45fd2829fc0aaad44facec9e2e74723/pkg/services/auth/auth_token.go#L208) as it has not yet seen the client coming with Token2
  - This time it [keeps Token1 in prev_auth_token](https://github.com/grafana/grafana/blob/299a0e20f45fd2829fc0aaad44facec9e2e74723/pkg/services/auth/auth_token.go#L231) so Token2 is now lost.
  - Token3 is stored in the response and query processing begins (and will take a while)
  - The first query finishes and Token2 is finally sent to the client.
  - The next request is sent before Token3 arrives so it has Token2 which Grafana no longer knows about
  - The client gets a 401 and logs itself out by going to `/logout` which cleans the auth cookie

**Environment**:
- Grafana version: v6.3.3 (84d0a71)
- Data source type & version: InfluxDB / Postgres
- OS Grafana is installed on: CentOS 7
- User OS & Browser: Chrome, Chromium, Firefox
- Grafana plugins:
- Others:

**Comment by torkelo** [comment]
strange, no other reports like this. @marefr any thoughts? Worth investigating/trying replicate soon or wait until we have more reports like this.

**Comment by mstaalesen** [comment]
Not sure if it is 100% related to this. But after upgrading from 6.2.5 to 6.3.3, we see that dashboard-pc's showing grafana on screen (rotating between e.g. grafana and other dashboards from other tools), the user is logged out after ~12 hours.

**Comment by marefr** [comment]
When implementing the new auth token solution we initially wanted to do the lookup before handling the actual request and then do the rotation just after handling the request (before sending response to the client). Like this:
1. Request comes in
2. Lookup auth token
3. Call actual request handler
4. Try rotate auth token
5. Response to client

But due to technical issues with the web framework (go macaron) we couldn't get this to work and had to revert back to the following (current) solution:
1. Request comes in
2. Lookup auth token and try rotate auth token
3. Call actual request handler
4. Response to client

In summary, I don't think the future-proof solution is to increase the `urgentRotateTime` ([code](https://github.com/grafana/grafana/blob/master/pkg/services/auth/auth_token.go#L26)) since timeouts may differ from user to user so hard to set it low enough. Instead I would like to implement our initial planned solution. However, short term I think we'll maybe need to add introduce a setting for the `urgentRotateTime`.

**Comment by yesoreyeram** [comment]
@marefr - Is this available in 6.6.1 ? I confirm that the issue exists in 6.6.1.

**Comment by marefr** [comment]
@yesoreyeram yes it's in 6.6.1. Can you reproduce it or provide information that can help reproduce it?

**Comment by yesoreyeram** [comment]
Hi @marefr  I dont have exact steps to reproduce. But I can see the users are frequently redirected to logout page. At the same time it matches with the grafana logs where it says my data source quieres are timed out. Here azure log analytics is the data source i am referring. ( I have azure generic oauth as my login provider. Previously using LDAP also we had issuesm ) . Mostly this occurs on the tv screens where the dashboard are continuously active. Sometimes it is happening every 10 to 15 mins. Let me know if you need more details.

**Comment by torkelo** [comment]
Maybe they take too long and are then cancelled by the next refresh , and hitting the token and refresh issue with cancelled requests that Carl has a PR for

**Comment by marefr** [comment]
Yeah, we have a couple of fixes for upcoming v6.6.2, for example #22106 that @torkelo referred to.

@yesoreyeram maybe you can inspect the Grafana server log to verify that you don't see any other error, like database related? Can you tell us a bit of the dashboards running on tv screens, what refresh interval and how long does the queries take in average?

**Comment by yesoreyeram** [comment]
I don't see any other errors in grafana logs except data sources timeout.  I do have multiple data sources. Graphite, Prometheus, influx db, elasticsearch, cloud watch and mostly azure monitor. Most of the timeouts are azure log analytics.  There are 5 different dashboards across 12 tv screens and set to auto refresh 2 to 5 minutes. Those tv screens share 3 different credentials to login.  Sometimes even though the screens are not using azure monitor on its own its getting log out. AssuMing it was caused by the same user creds used in different screen.

**Comment by paltaa** [comment]
I did hit this issue in version 11.3.0, got kicked out every 10 minutes if had multiple parallel queries in different browser tabs even if the /rotate endpoint responds with 200
